Lib/test/test_threading.py

```diff
@@ -256,32 +254,60 @@ def test_various_ops_large_stack(self):
 
     def test_foreign_thread(self):
         # Check that a "foreign" thread can use the threading module.
+        dummy_thread = None
+        error_message = None
         def f(mutex):
             # Calling current_thread() forces an entry for the foreign
             # thread to get made in the threading._active map.
-            threading.current_thread()
-            mutex.release()
+            try:
+                nonlocal dummy_thread
+                nonlocal error_message
+                dummy_thread = threading.current_thread()
+                tid = dummy_thread.ident
+                def verify_conditions():
+                    if tid not in threading._active:
+                        return f'Expected {tid} to be in {threading._active}'
+                    if not isinstance(dummy_thread, threading._DummyThread):
+                        return f'Expected {dummy_thread} to be an instance of threading._DummyThread.'
+                    found = threading._active.get(tid)
+                    if found is not dummy_thread:
+                        return f'Expected {found} to be {dummy_thread}.'
+                    if not dummy_thread.is_alive():
+                        # gh-29376
+                        return 'Expected _DummyThread to be considered alive.'
+                    r = repr(threading._active[tid])
+                    if '_DummyThread' not in r:
+                        return f'Expected "_DummyThread" to be in {r}.'
+                    return None
+                error_message = verify_conditions()
+            finally:
+                mutex.release()
 
         mutex = threading.Lock()
         mutex.acquire()
         with threading_helper.wait_threads_exit():
             tid = _thread.start_new_thread(f, (mutex,))
             # Wait for the thread to finish.
             mutex.acquire()
-        self.assertIn(tid, threading._active)
-        self.assertIsInstance(threading._active[tid], threading._DummyThread)
-        #Issue 29376
-        self.assertTrue(threading._active[tid].is_alive())
-        self.assertRegex(repr(threading._active[tid]), '_DummyThread')
-
+        if error_message is not None:
+            self.fail(error_message)
+        self.assertEqual(tid, dummy_thread.ident)
         # Issue gh-106236:
         with self.assertRaises(RuntimeError):
-            threading._active[tid].join()
-        threading._active[tid]._started.clear()
+            dummy_thread.join()
+        dummy_thread._started.clear()
         with self.assertRaises(RuntimeError):
-            threading._active[tid].is_alive()
-
-        del threading._active[tid]
+            dummy_thread.is_alive()
+        # Busy wait for the following condition: after the thread dies, the
```

Comment:
Do we need to call the garbage collector here?

Reply:
Well, I checked that the test is passed with disabled GC. So there should not be issue on PyPy.

Reply:
The problem with PyPy is that it'll require GC to kick in to collect that object to then remove it from '_active`, whereas on Python the deallocation time is "guaranteed" to be as soon as the thread is collected as reference counting is more deterministic.

So, it's not that on PyPy this will raise an error or anything like that, more something like: there'll be a delay between the thread being collected and being removed from `_active`.